app/scripts/background.js

```diff
@@ -91,16 +92,60 @@ const PHISHING_WARNING_PAGE_TIMEOUT = ONE_SECOND_IN_MILLISECONDS;
 /**
  * In case of MV3 we attach a "onConnect" event listener as soon as the application is initialised.
  * Reason is that in case of MV3 a delay in doing this was resulting in missing first connect event after service worker is re-activated.
+ *
+ * @param remotePort
  */
-
 const initApp = async (remotePort) => {
   browser.runtime.onConnect.removeListener(initApp);
   await initialize(remotePort);
   log.info('MetaMask initialization complete.');
 };
 
+const throwErrorIfLastErrorFound = () => {
+  const err = checkForError();
+  if (err) {
+    throw err;
+  }
+};
+
+/**
+ * Sends a message to the dapp(s) content script to signal it can connect to MetaMask background as
+ * the backend is not active. It is required to re-connect dapps after service worker re-activates.
+ * For non-dapp pages, the message will be sent and ignored.
+ */
+const sendReadyMessageToTabs = async () => {
+  try {
+    const tabs = await browser.tabs.query({});
+    throwErrorIfLastErrorFound();
+
+    /* TODO we should only sendMessage to dapp tabs, not all tabs. */
+    for (const tab of tabs) {
+      browser.tabs
+        .sendMessage(tab.id, {
+          name: EXTENSION_MESSAGES.READY,
+        })
+        .catch((e) => {
+          if (
+            e.message ===
+            'Could not establish connection. Receiving end does not exist.'
+          ) {
+            /** Safely ignore this error, as it means the tab may not be a dapp. */
+          } else {
+            throw new Error(e);
+          }
+        })
+        .finally(() => {
+          throwErrorIfLastErrorFound();
```

Comment:
May be we can just log error here, we do not need to re-throw.

Reply:
@jpuri 

Good ideas. I created a new method, `checkForErrorAndLog` in the `metamask-extension/app/scripts/lib/util.js` file to use. 

I also removed the catch statement here. I discovered that the "runtime.lastError: Could not establish connection." error was being thrown here because we were querying all tabs, including tabs without URLs, e.g., devTool tabs like `chrome://extensions`. These tabs were unable to receive the messages. 

With the latest changes, I'm no longer seeing the "lastError: Could not establish connection." error on the background page